src/index.js

```diff
@@ -58,26 +68,32 @@ export let ContextTree = _.curry(
     )
 
     // overwriting extend
-    extend = _.over([extend, onChange])
+    extend = _.over([extend, (a, b) => TreeInstance.onChange(a, b)])
 
     // Getting the Traversals
     let { markForUpdate, markLastUpdate, prepForUpdate } = traversals(extend)
 
     let processEvent = event => path =>
-      F.flurry(
+      _.flow(
         getAffectedNodes(customReactors, getNode, types),
         // Mark children only if it's not a parent of the target so we don't incorrectly mark siblings
-        _.each(n => {
-          F.unless(isParent(n.path, event.path), Tree.walk)(markForUpdate)(n)
-        })
+        _.flatMap((
+          n // flatMap because traversing children can create arrays
+        ) =>
+          F.unless(isParent(n.path, event.path), Tree.toArrayBy)(markForUpdate)(
+            n
+          )
+        )
       )(event, path)
 
     // Event Handling
     let dispatch = async event => {
       log(`${event.type} event at ${event.path}`)
       await validate(runTypeFunction(types, 'validate'), extend, tree)
-      bubbleUp(processEvent(event), event.path)
+      let updatedNodes = _.flatten(bubbleUp(processEvent(event), event.path))
+      await Promise.all(_.invokeMap('onMarkForUpdate', updatedNodes))
```

Comment:
should this be in random order `all` or in sequential order `each` ?

Reply:
order doesn't matter